File: yoga-bot-whatsapp/src/services/chatService.js
const http = require('http');
const { URL } = require('url');
const { v4: uuidv4 } = require('uuid');
const config = require('../config/config');
const whatsappService = require('./whatsappService');
const ttsService = require('./ttsService');
const { processMarkdown, extractProductInfo } = require('../utils/textUtils');

class ChatService {
    constructor() {
        this.sessionMap = new Map(); // senderId => sessionId
        this.historyMap = new Map(); // senderId => [{ user, bot }]
        this.languageMap = new Map(); // senderId => language code
    }

    async handleChat(userMessage, userNumber, senderId) {
        const sessionId = this.sessionMap.get(senderId);
        const language = this.languageMap.get(senderId);
        const apiUrl = new URL(`${config.API_BASE_URL}/v1/chat/message`);
        apiUrl.searchParams.append('message', userMessage);
        apiUrl.searchParams.append('session_id', sessionId);
        
        const options = {
            method: 'POST',
            headers: {
                Accept: 'text/event-stream',
            },
        };

        const req = http.request(apiUrl, options, async res => {
            let fullText = '';
            res.on('data', chunk => {
                fullText += chunk.toString();
            });

            res.on('end', async () => {
                if (!fullText) {
                    await whatsappService.sendMessage(userNumber, '⚠️ No response from backend.');
                    return;
                }

                const { products, remainingText } = extractProductInfo(fullText, language);
                
                if (products.length > 0) {
                    await this._handleProductResponse(products, remainingText, userNumber, language);
                } else {
                    await this._handleRegularResponse(fullText, userMessage, userNumber, senderId, language);
                }
            });
        });

        req.on('error', async err => {
            console.error('❗ API request failed:', err);
            await whatsappService.sendMessage(userNumber, '🚨 Failed to connect to backend.');
        });

        req.end();
    }

    async _handleProductResponse(products, remainingText, userNumber, language) {
        const introText = remainingText.split('\n\n')[0] + '\n\n';
        await whatsappService.sendMessage(userNumber, introText);

        for (const product of products) {
            try {
                const response = await fetch(product.imageUrl);
                const imageBuffer = await response.buffer();
                
                const priceLabel = language === 'hin' ? 'कीमत' : 'Price';
                const caption = `*${product.name}*\n${priceLabel}: ₹${product.price}\n${product.link}`;
                
                await whatsappService.sendImage(userNumber, imageBuffer, caption);
            } catch (error) {
                console.error('Error sending product image:', error);
                const priceLabel = language === 'hin' ? 'कीमत' : 'Price';
                await whatsappService.sendMessage(userNumber, `*${product.name}*\n${priceLabel}: ₹${product.price}\n${product.link}`);
            }
        }

        let closingText = remainingText;
        if (!closingText) {
            closingText = language === 'hin' 
                ? "\nक्या आप इन विकल्पों के बारे में अधिक जानकारी चाहते हैं या कुछ और सहायता चाहिए?\n\n👉 मुख्य मेनू के लिए *M* टाइप करें"
                : "\nWould you like more information on these options or assistance with something else?\n\n👉 Type *M* for Main Menu";
        } else {
            const menuOption = language === 'hin'
                ? "\n\n👉 मुख्य मेनू के लिए *M* टाइप करें"
                : "\n\n👉 Type *M* for Main Menu";
            closingText += menuOption;
        }

        closingText = processMarkdown(closingText);
        await whatsappService.sendMessage(userNumber, closingText);
    }

    async _handleRegularResponse(fullText, userMessage, userNumber, senderId, language) {
        const formattedText = processMarkdown(fullText);

        // Save to history
        const history = this.historyMap.get(senderId) || [];
        history.push({ user: userMessage, bot: fullText });
        this.historyMap.set(senderId, history);

        try {
            const reply = `${formattedText}\n\n👉 Type *M* for Main Menu`;
            await whatsappService.sendMessage(userNumber, reply.slice(0, 4000));

            const audioFile = await ttsService.getTTSAudio(fullText, language);
            const oggFile = await ttsService.convertToOgg(audioFile);
            await whatsappService.sendAudio(userNumber, oggFile);
            
            // Clean up audio files
            fs.unlinkSync(audioFile);
            fs.unlinkSync(oggFile);
        } catch (error) {
            console.error('Error generating audio response:', error);
        }
    }

    async showLanguageMenu(userNumber) {
        let menuText = "Namaste!\n";
        menuText += "नमस्ते!\n\n";
        
        menuText += "Please choose your preferred language:\n";
        menuText += "कृपया अपनी पसंदीदा भाषा चुनें:\n\n";
        
        Object.entries(config.SUPPORTED_LANGUAGES).forEach(([code, lang]) => {
            const hindiNames = {
                'eng': 'English',
                'hin': 'हिंदी'
            };
            menuText += `•  ${code} – ${hindiNames[code]}\n`;
        });
        
        menuText += "\n👉 Type the language code to continue (e.g., eng for English)\n";
        menuText += "👉 जारी रखने के लिए भाषा कोड टाइप करें (जैसे, hin हिंदी के लिए)";
        
        await whatsappService.sendMessage(userNumber, menuText);
    }

    setLanguage(senderId, language) {
        this.languageMap.set(senderId, language);
        if (!this.sessionMap.has(senderId)) {
            this.sessionMap.set(senderId, uuidv4());
        }
    }

    startNewSession(senderId) {
        const newSession = uuidv4();
        this.sessionMap.set(senderId, newSession);
        this.historyMap.set(senderId, []);
        this.languageMap.delete(senderId);
    }

    getHistory(senderId) {
        return this.historyMap.get(senderId) || [];
    }

    getLanguage(senderId) {
        return this.languageMap.get(senderId);
    }

    hasLanguage(senderId) {
        return this.languageMap.has(senderId);
    }
}

module.exports = new ChatService(); 

File: yoga-bot-web/app/services/chat_service.py
import uuid
from typing import Literal

import httpx
from app.config import get_settings
from app.utils.text_utils import extract_product_info, process_markdown

LanguageCode = Literal["eng", "hin"]

MENU_OPTIONS = {
    "eng": {
        "startNewSession": "Start New Session",
        "history": "View Chat History",
        "changeLanguage": "Change Language",
        "noHistory": "📭 No history available for this session.",
        "historyTitle": "🕘 *Chat History:*\n",
        "youSaid": "🗣 *You said:*",
        "botName": "*YogaBot:*",
        "userPrefix": "*You:*",
        "mainMenu": "📋 Main Menu",
        "options": "👉 *Options:*\n1. Start New Session\n2. View Chat History\n3. Change Language",
        "menuSuffix": "\n\n👉 Type *M* for Main Menu",
        "productClosing": "\nWould you like more information on these options or assistance with something else?\n\n👉 Type *M* for Main Menu",
    },
    "hin": {
        "startNewSession": "नया सत्र शुरू करें",
        "history": "चैट इतिहास देखें",
        "changeLanguage": "भाषा बदलें",
        "noHistory": "📭 इस सत्र के लिए कोई इतिहास उपलब्ध नहीं है।",
        "historyTitle": "🕘 *चैट इतिहास:*\n",
        "youSaid": "🗣 *आपने कहा:*",
        "botName": "*योगबॉट:*",
        "userPrefix": "*आप:*",
        "mainMenu": "📋 मुख्य मेनू",
        "options": "👉 *विकल्प:*\n1. नया सत्र शुरू करें\n2. चैट इतिहास देखें\n3. भाषा बदलें",
        "menuSuffix": "\n\n👉 मुख्य मेनू के लिए *M* टाइप करें",
        "productClosing": "\nक्या आप इन विकल्पों के बारे में अधिक जानकारी चाहते हैं या कुछ और सहायता चाहिए?\n\n👉 मुख्य मेनू के लिए *M* टाइप करें",
    },
}


class ChatService:
    def __init__(self):
        self._session_map: dict[str, str] = {}
        self._history_map: dict[str, list[dict]] = {}
        self._language_map: dict[str, str] = {}

    def _get_or_create_session(self, client_id: str) -> str:
        if client_id not in self._session_map:
            self._session_map[client_id] = str(uuid.uuid4())
        return self._session_map[client_id]

    def set_language(self, client_id: str, language: LanguageCode) -> None:
        self._language_map[client_id] = language
        self._get_or_create_session(client_id)

    def get_language(self, client_id: str) -> LanguageCode | None:
        return self._language_map.get(client_id)

    def has_language(self, client_id: str) -> bool:
        return client_id in self._language_map

    def start_new_session(self, client_id: str) -> None:
        self._session_map[client_id] = str(uuid.uuid4())
        self._history_map[client_id] = []
        self._language_map.pop(client_id, None)

    def get_history(self, client_id: str) -> list[dict]:
        return self._history_map.get(client_id, [])

    def get_session_id(self, client_id: str) -> str | None:
        return self._session_map.get(client_id)

    async def call_rag(self, message: str, session_id: str) -> str:
        settings = get_settings()
        url = f"{settings.API_BASE_URL.rstrip('/')}/v1/chat/message"
        params = {"message": message, "session_id": session_id}
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.get(
                url, params=params, headers={"Accept": "text/event-stream"}
            )
            response.raise_for_status()
            return response.text

    async def handle_chat(
        self, user_message: str, client_id: str, language: LanguageCode
    ) -> dict:
        """Returns { "text", "products"?, "audio_base64"? }."""
        session_id = self._get_or_create_session(client_id)
        raw = await self.call_rag(user_message, session_id)
        if not raw or not raw.strip():
            return {"text": "⚠️ No response from backend.", "products": None}

        products, remaining_text = extract_product_info(raw, language)
        menu = MENU_OPTIONS[language]

        if products:
            intro = remaining_text.split("\n\n")[0] + "\n\n" if remaining_text else ""
            closing = remaining_text[len(intro) :].strip() if remaining_text else ""
            if not closing:
                closing = menu["productClosing"]
            else:
                closing += menu["menuSuffix"]
            closing = process_markdown(closing)
            return {
                "text": intro + closing,
                "products": [
                    {
                        "name": p["name"],
                        "price": p["price"],
                        "link": p["link"],
                        "imageUrl": p["imageUrl"],
                    }
                    for p in products
                ],
            }

        formatted = process_markdown(raw)
        history = self._history_map.get(client_id, [])
        history.append({"user": user_message, "bot": raw})
        self._history_map[client_id] = history

        reply_text = formatted + menu["menuSuffix"]
        result = {"text": reply_text[:4000], "products": None}

        # Optional: add TTS in background or via separate endpoint
        try:
            from app.services.tts_service import get_tts_base64
            result["audio_base64"] = await get_tts_base64(raw, language)
        except Exception:
            pass
        return result

    def get_menu_options(self, language: LanguageCode) -> dict:
        return MENU_OPTIONS.get(language, MENU_OPTIONS["eng"])


def get_language_menu_message() -> tuple[str, list[str]]:
    msg = (
        "Namaste!\nनमस्ते!\n\n"
        "Please choose your preferred language:\n"
        "कृपया अपनी पसंदीदा भाषा चुनें:\n\n"
        "• eng – English\n• hin – हिंदी\n\n"
        "👉 Type the language code to continue (e.g., eng or hin)"
    )
    return msg, ["eng", "hin"]


chat_service = ChatService()
